flink-core/src/test/java/org/apache/flink/api/connector/source/lib/DataGeneratorSourceTest.java

```diff
@@ -0,0 +1,250 @@
+                } else {
+                    reader.addSplits(splits);
+                }
+            }
+        }
+
+        final List<Long> result = out.getEmittedRecords();
+        validateSequence(result, from, to);
+    }
+
+    private static void validateSequence(
+            final List<Long> sequence, final long from, final long to) {
+        if (sequence.size() != to - from + 1) {
+            failSequence(sequence, from, to);
+        }
+
+        long nextExpected = from;
+        for (Long next : sequence) {
+            if (next != nextExpected++) {
+                failSequence(sequence, from, to);
+            }
+        }
+    }
+
+    private static void failSequence(final List<Long> sequence, final long from, final long to) {
+        Assertions.fail(
+                String.format(
+                        "Expected: A sequence [%d, %d], but found: sequence (size %d) : %s",
+                        from, to, sequence.size(), sequence));
+    }
+
+    private static SourceReader<Long, NumberSequenceSource.NumberSequenceSplit> createReader()
+            throws Exception {
+        // the arguments passed in the source constructor matter only to the enumerator
+        GeneratorFunction<Long, Long> generatorFunctionStateless = index -> index;
+        DataGeneratorSource<Long> dataGeneratorSource =
+                new DataGeneratorSource<>(generatorFunctionStateless, Long.MAX_VALUE, Types.LONG);
+
+        return dataGeneratorSource.createReader(new DummyReaderContext());
+    }
+
+    // ------------------------------------------------------------------------
+    //  test utils / mocks
+    //
+    //  the "flink-connector-test-utils module has proper mocks and utils,
+    //  but cannot be used here, because it would create a cyclic dependency.
+    // ------------------------------------------------------------------------
+
+    private static final class DummyReaderContext implements SourceReaderContext {
+
+        @Override
+        public SourceReaderMetricGroup metricGroup() {
+            return UnregisteredMetricsGroup.createSourceReaderMetricGroup();
+        }
+
+        @Override
+        public Configuration getConfiguration() {
+            return new Configuration();
+        }
+
+        @Override
+        public String getLocalHostName() {
+            return "localhost";
+        }
+
+        @Override
+        public int getIndexOfSubtask() {
+            return 0;
+        }
+
+        @Override
+        public void sendSplitRequest() {}
+
+        @Override
+        public void sendSourceEventToCoordinator(SourceEvent sourceEvent) {}
+
+        @Override
+        public UserCodeClassLoader getUserCodeClassLoader() {
+            return SimpleUserCodeClassLoader.create(getClass().getClassLoader());
+        }
+
+        @Override
+        public int currentParallelism() {
+            return 1;
+        }
+    }
+
+    private static final class TestingReaderOutput<E> implements ReaderOutput<E> {
```

Comment:
Hmm... good question. This PR should enable us to deprecate the `SourceFunction`. My understanding is that we would want to provide alternatives for the respective high-level methods that are currently available in the environment:
https://github.com/apache/flink/blob/eb69b11532e6ac3fafd74849ed5f8a2c1273b71c/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java#L1142
https://github.com/apache/flink/blob/eb69b11532e6ac3fafd74849ed5f8a2c1273b71c/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java#L1296-L1300

Reply:
Indeed, but that should still be possible by having flink-streaming-java depend on that module.
At first I was worried about a circular dependency w.r.t. testing, but since the connector currently lives in flink-core this shouldn't be a problem.

Reply:
We have the choice here to either keep the rate-limiting related classes in core or also move them along into the datagen connector. To me, they seem generic enough to belong in core and I see them being potentially useful elsewhere, but I understand that you prefer (and for good reasons) to keep the surface of exposed functionality to only what is supported by concrete demand. What is your preference?

Reply:
I'm tempted to keep it in flink-core because that's where the Source interface is, and I feel like outside of the source the rate limiter would be difficult/weird to use due to it's async nature.